swarm/storage/pushsync/pusher.go

```diff
@@ -0,0 +1,277 @@
+		pushed:   make(map[string]*pushedItem),
+		receipts: make(chan chunk.Address),
+		ps:       ps,
+	}
+	go p.sync()
+	return p
+}
+
+// Close closes the pusher
+func (p *Pusher) Close() {
+	close(p.quit)
+}
+
+// sync starts a forever loop that pushes chunks to their neighbourhood
+// and receives receipts (statements of custody) for them.
+// chunks that are not acknowledged with a receipt are retried
+// not earlier than retryInterval after they were last pushed
+// the routine also updates counts of states on a tag in order
+// to monitor the proportion of saved, sent and synced chunks of
+// a file or collection
+func (p *Pusher) sync() {
+	var chunks <-chan chunk.Chunk
+	var cancel, stop func()
+	var ctx context.Context
+	var synced []storage.Address
+
+	// timer
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	// register handler for pssReceiptTopic on pss pubsub
+	deregister := p.ps.Register(pssReceiptTopic, false, func(msg []byte, _ *p2p.Peer) error {
+		return p.handleReceiptMsg(msg)
+	})
+	defer deregister()
+
+	chunksInBatch := -1
+	var batchStartTime time.Time
+
+	for {
+		select {
+
+		// retry interval timer triggers starting from new
+		case <-timer.C:
+			metrics.GetOrRegisterCounter("pusher.subscribe-push", nil).Inc(1)
+			// TODO: implement some smart retry strategy relying on sent/synced ratio change
+			// if subscribe was running, stop it
+			if stop != nil {
+				stop()
+			}
+			for _, addr := range synced {
+				// set chunk status to synced, insert to db GC index
+				if err := p.store.Set(context.Background(), chunk.ModeSetSync, addr); err != nil {
+					log.Warn("error setting chunk to synced", "addr", addr, "err", err)
+					continue
+				}
+				delete(p.pushed, addr.Hex())
+			}
+			// reset synced list
+			synced = nil
+
+			// we don't want to record the first iteration
+			if chunksInBatch != -1 {
+				// this measurement is not a timer, but we want a histogram, so it fits the data structure
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.hist", nil).Update(time.Duration(chunksInBatch))
+
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.time", nil).UpdateSince(batchStartTime)
+				metrics.GetOrRegisterCounter("pusher.subscribe-push.chunks-in-batch", nil).Inc(int64(chunksInBatch))
+			}
+			chunksInBatch = 0
+			batchStartTime = time.Now()
+
+			// and start iterating on Push index from the beginning
+			ctx, cancel = context.WithCancel(context.Background())
+			chunks, stop = p.store.SubscribePush(ctx)
+			// reset timer to go off after retryInterval
+			timer.Reset(retryInterval)
+
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			chunksInBatch++
+			// if no more, set to nil and wait for timer
+			if !more {
+				chunks = nil
+				continue
+			}
+
+			metrics.GetOrRegisterCounter("pusher.send-chunk", nil).Inc(1)
+			// if no need to sync this chunk then continue
+			if !p.needToSync(ch) {
+				continue
```

Comment:
Yes, if you continue, it means that the chunk doesn't need to be synced, so it has been processed and once we got a receipt back, it will be marked for GC.